controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -1164,6 +1164,20 @@ func (r *BareMetalHostReconciler) actionPreparing(prov provisioner.Provisioner,
 		return recordActionFailure(info, metal3api.PreparationError, provResult.ErrorMessage)
 	}
 
+	if hfcDirty && started {
+		hfcStillDirty, err := r.saveHostFirmwareComponents(prov, info, hfc)
```

Comment:
if there is a failure we shouldn't retry I would say, it can be a bad firmware (Dell for example has firmware separate for each model, if I use the firmware of an R750 in R640 it complains and fails)

Reply:
Or it could be a dropped connection or a failed write.
We shouldn't report that we've done something if we haven't done it.
If the user provides the wrong firmware we should keep trying until they realise and stop doing that.

Reply:
Yeah, unless we invent a way for Ironic to say "this will never work, don't try again", the current trend is to retry.

Reply:
```
	if provResult.ErrorMessage != "" {
		if bmhDirty {
			info.log.Info("handling cleaning error in controller")
			clearHostProvisioningSettings(info.host)
		}
		if hfcDitry {
			clearHostFirmwareComponentsUpdates(hfc)
		}
		return recordActionFailure(info, metal3api.PreparationError, provResult.ErrorMessage)
	}
```
Something like this?